include/fsm/fsm.h

```diff
@@ -32,6 +32,13 @@ typedef unsigned int fsm_state_t;
  * their end states. These can be used to determine which of the
  * original FSM(s) matched when executing a combined FSM. */
 typedef unsigned int fsm_end_id_t;
+#define FSM_END_ID_NONE ((unsigned int)-1)
+
+/* struct used to return a collection of end IDs. */
+struct fsm_end_ids {
+	unsigned count;
+	fsm_end_id_t ids[1];
+};
```

Comment:
Did you introduce this because the set/adt stuff isn't exposed to users? That's perfect, thanks

Reply:
Right, this represents opaque numeric IDs associated with endpoints, but the internals know how to update them across operations. The caller can focus on what metadata is associated with them, but no longer needs a `carryopaque`-style interface that is coupled to the implementation details.